File: Homework3/src/main/java/Tinkoff/Address.java
package Tinkoff;

class Address {

    public Address(String postcode, String country,
                   String region, String city,
                   String street, String house, String apartment) {
        this.postcode = postcode;
        this.country = country;
        this.region = region;
        this.city = city;
        this.street = street;
        this.house = house;
        this.apartment = apartment;
    }
    @Override
    public String toString()
    {
        return postcode+", "+country+", "+region+", "+city+", "+street+", "+house+", "+apartment;
    }
    
    private final String postcode;
    private final String country;
    private final String region;
    private final String city;
    private final String street;
    private final String house;
    private final String apartment;
}

File: homeworkScalaTwo/src/main/scala/Address.scala
object Address {

  case class Address(street: String, house: Int, apartment: Option[Int]) {
    override def toString: String = s"$street $house ${apartment.getOrElse("")}"

  }
  def isEmpty(x: String): Boolean = x == null || x.isEmpty

  implicit def listToAddress(list: List[Any]): Address ={
    if (list.nonEmpty && list.head.toString==""){
      throw new Exception("Can't construct Address from")
    }
    list match {
    case List(street: String, house: Int, apartment: Int, _*) => Address(street, house, Some(apartment))
    case List(street: String, house: Int, apartment: Option[_], _*) =>
      apartment match {
        case None => Address(street, house, None)
        case Some(apartment: Int) => Address(street, house, Some(apartment))
        case _=>throw new Exception("Can't construct Address from")
      }
    case List(street: String, house: Int, _*) => Address(street, house, None)
    case List(_) => throw new Exception("Can't construct Address from")
  }}

  def printAddress(address: Address): Unit = {
    println(address.toString)
  }

  def main(args: Array[String]): Unit = {
    printAddress(List("Профсоюзная", 83, Some(1)))
    printAddress(List("Профсоюзная", 83, 1))
    printAddress(List("Профсоюзная", 83))
    printAddress(List("Профсоюзная", 83, 1, "A", "B"))
    printAddress(List("Профсоюзная", 83, "C"))
    printAddress(List("Профсоюзная"))
  }
}
